--- 1.6.5/fs/tomoyo_audit.c	2008-11-11 09:00:00.000000000 +0900
+++ 1.6.5-hotfix/fs/tomoyo_audit.c	2008-12-09 20:55:52.000000000 +0900
@@ -244,7 +244,7 @@ int ccs_write_audit_log(const bool is_gr
 	struct task_struct *task = current;
 	if (!r->domain)
 		r->domain = task->domain_info;
-	if (ccs_can_save_audit_log(r->domain, is_granted) < 0)
+	if (!ccs_can_save_audit_log(r->domain, is_granted))
 		goto out;
 	va_start(args, fmt);
 	len = vsnprintf((char *) &pos, sizeof(pos) - 1, fmt, args) + 32;
