Signed-off-by: Kentaro Takeda <takedakn@nttdata.co.jp>
Signed-off-by: Tetsuo Handa <penguin-kernel@I-love.SAKURA.ne.jp>

---
 security/tomoyo/audit.c |   54 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 54 insertions(+)

Index: src/security/tomoyo/audit.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ src/security/tomoyo/audit.c	2007-08-07 17:32:13.000000000 +0900
@@ -0,0 +1,54 @@
+/*
+ * security/tomoyo/audit.c
+ *
+ * Audit functions for TOMOYO Linux
+ */
+
+#include "tomoyo.h"
+#include <linux/audit.h>
+
+/* move to include/linux/audit.h */
+#define AUDIT_TMY_GRANTED 2001
+#define AUDIT_TMY_REJECTED 2002
+
+char *tmy_init_audit_log(int *len)
+{
+	char *buf;
+	struct task_struct *task = current;
+	const char *domainname = TMY_SECURITY->domain->domainname->name;
+
+	*len += strlen(domainname) + 256;
+	buf = tmy_alloc(*len);
+
+	if (buf != NULL)
+		snprintf(buf,
+			 (*len) - 1,
+			 "pid=%d uid=%d gid=%d euid=%d egid=%d "
+			 "suid=%d sgid=%d fsuid=%d fsgid=%d domain=\"%s\" ",
+			 task->pid, task->uid,
+			 task->gid, task->euid, task->egid,
+			 task->suid, task->sgid, task->fsuid,
+			 task->fsgid, domainname);
+
+	return buf;
+}
+
+/*
+ * Write audit log.
+ * Caller must allocate buf with tmy_init_audit_log().
+ */
+int tmy_write_audit_log(char *buf, const int is_granted, const int is_enforce)
+{
+	struct audit_buffer *ab;
+	int type = is_granted ? AUDIT_TMY_GRANTED : AUDIT_TMY_REJECTED;
+
+	ab = audit_log_start(current->audit_context, GFP_KERNEL, type);
+	if (ab) {
+		const char *msg = is_granted ? "GRANTED" : tmy_get_msg(is_enforce);
+		audit_log_format(ab, "TOMOYO-%s: %s", msg, buf);
+		audit_log_end(ab);
+	}
+
+	tmy_free(buf);
+	return ab ? 0 : -ENOMEM;
+}
