diff -urp 1.6.4/fs/tomoyo_cond.c 1.6.4-hotfix/fs/tomoyo_cond.c
--- 1.6.4/fs/tomoyo_cond.c	2008-09-03 00:00:00.000000000 +0900
+++ 1.6.4-hotfix/fs/tomoyo_cond.c	2008-10-14 16:38:51.000000000 +0900
@@ -1031,8 +1031,8 @@ bool ccs_check_condition(const struct ac
 		const u8 left = header >> 8;
 		const u8 right = header;
 		ptr++;
-		if ((left >= PATH1_UID && left < MAX_KEYWORD) ||
-		    (right >= PATH1_UID && right < MAX_KEYWORD)) {
+		if ((left >= PATH1_UID && left < EXEC_ARGC) ||
+		    (right >= PATH1_UID && right < EXEC_ARGC)) {
 			if (!obj)
 				goto out;
 			if (!obj->validate_done) {
