--- ccs-tools.spec
+++ ccs-tools.spec
@@ -8,12 +8,7 @@
 ExclusiveOS: Linux
 Autoreqprov: no
 Buildroot: %{_tmppath}/%{name}-%{version}-%{release}-buildroot
-##
-## This spec file is intended for distribution independent.
-## I don't enable "BuildRequires:" line because rpmbuild will fail on
-## environments where packages are managed by (e.g.) apt.
-##
-# BuildRequires: ncurses-devel
+BuildRequires: ncurses-devel
 Requires: ncurses
 Conflicts: ccs-tools < 1.8.3-2
